### ISSUE TITLE
Partitions unavailable due to broken gateway topology

### ISSUE THREAD

**Issue by npepinpe** [zeebe-io/zeebe#4557]
**Describe the bug**

It seems possible sometimes for a gateway to have the wrong topology - it is unclear if it is missing topology updates or if there is a bug when merging these updates together, but essentially we ran into the case where we had a 3 node cluster, 4 partitions, replication factor 3, and zeebe-2's gateway topology did not know the leader for partition 2 and 3, which was zeebe-1. Furthermore, it thought zeebe-1 only had two partitions: partition 1 and 4, which may indicate there was a conflict.

**To Reproduce**

This often happens in the cloud end to end tests.

**Expected behavior**

The topology of every gateway is eventually consistent, such that we can route requests to the correct partitions. At the moment, it sometimes breaks and never recovers until a new failover occurs.

**Log/Stacktrace**

Immi will attach the logs and the data partition soon.

**Environment:**
 - OS: k8s
 - Zeebe Version: 0.23.1
 - Configuration: to be filled by Immi

**Comment by hisImminence** [comment]
**Logs of brokers topology:**
**kubectl exec -it zeebe-2 -n 6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe -- zbctl --insecure status**
Cluster size: 3
Partitions count: 4
Replication factor: 3
Gateway version: 0.23.1
Brokers:
  Broker 0 - zeebe-0.zeebe-broker-service.6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe.svc.cluster.local:26501
    Version: 0.23.1
    Partition 1 : Follower
    Partition 2 : Follower
    Partition 3 : Follower
    Partition 4 : Follower
  Broker 1 - zeebe-1.zeebe-broker-service.6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe.svc.cluster.local:26501
    Version: 0.23.1
    Partition 1 : Follower
    Partition 4 : Follower
  Broker 2 - zeebe-2.zeebe-broker-service.6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe.svc.cluster.local:26501
    Version: 0.23.1
    Partition 1 : Leader
    Partition 2 : Follower
    Partition 3 : Follower
    Partition 4 : Leader
➜  ~ (⎈ cc_europe-west1-d_ultratest:default)
**kubectl exec -it zeebe-1 -n 6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe -- zbctl --insecure status**
Cluster size: 3
Partitions count: 4
Replication factor: 3
Gateway version: 0.23.1
Brokers:
  Broker 0 - zeebe-0.zeebe-broker-service.6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe.svc.cluster.local:26501
    Version: 0.23.1
    Partition 1 : Follower
    Partition 2 : Follower
    Partition 3 : Follower
    Partition 4 : Follower
  Broker 1 - zeebe-1.zeebe-broker-service.6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe.svc.cluster.local:26501
    Version: 0.23.1
    Partition 1 : Follower
    Partition 2 : Leader
    Partition 3 : Leader
    Partition 4 : Follower
  Broker 2 - zeebe-2.zeebe-broker-service.6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe.svc.cluster.local:26501
    Version: 0.23.1
    Partition 1 : Leader
    Partition 2 : Follower
    Partition 3 : Follower
    Partition 4 : Leader
➜  ~ (⎈ cc_europe-west1-d_ultratest:default)
**kubectl exec -it zeebe-0 -n 6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe -- zbctl --insecure status**
Cluster size: 3
Partitions count: 4
Replication factor: 3
Gateway version: 0.23.1
Brokers:
  Broker 0 - zeebe-0.zeebe-broker-service.6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe.svc.cluster.local:26501
    Version: 0.23.1
    Partition 1 : Follower
    Partition 2 : Follower
    Partition 3 : Follower
    Partition 4 : Follower
  Broker 1 - zeebe-1.zeebe-broker-service.6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe.svc.cluster.local:26501
    Version: 0.23.1
    Partition 1 : Follower
    Partition 2 : Leader
    Partition 3 : Leader
    Partition 4 : Follower
  Broker 2 - zeebe-2.zeebe-broker-service.6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe.svc.cluster.local:26501
    Version: 0.23.1
    Partition 1 : Leader
    Partition 2 : Follower
    Partition 3 : Follower
    Partition 4 : Leader

**Comment by hisImminence** [comment]
**k -n 6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe get pod zeebe-0 -o yaml**
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: "2020-05-19T11:35:02Z"
  deletionGracePeriodSeconds: 30
  deletionTimestamp: "2020-05-19T13:12:55Z"
  generateName: zeebe-
  labels:
    app.kubernetes.io/app: zeebe
    app.kubernetes.io/component: gateway
    controller-revision-hash: zeebe-798cbcbbb4
    statefulset.kubernetes.io/pod-name: zeebe-0
  name: zeebe-0
  namespace: 6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: StatefulSet
    name: zeebe
    uid: be7cb84b-6b98-4122-b88b-fd820be7bad6
  resourceVersion: "147406700"
  selfLink: /api/v1/namespaces/6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe/pods/zeebe-0
  uid: 6b54268d-a9bb-4271-8ded-1eb92cb69ce8
spec:
  containers:
  - env:
    - name: JAVA_TOOL_OPTIONS
      value: -XX:MaxRAMPercentage=50.0 -XX:InitialRAMPercentage=25.0
    - name: K8S_NAMESPACE
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.namespace
    - name: K8S_POD_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    - name: ZEEBE_BROKER_CLUSTER_CLUSTERSIZE
      value: "3"
    - name: ZEEBE_BROKER_CLUSTER_PARTITIONSCOUNT
      value: "4"
    - name: ZEEBE_BROKER_CLUSTER_REPLICATIONFACTOR
      value: "3"
    - name: ZEEBE_BROKER_DATA_SNAPSHOTPERIOD
      value: 5m
    - name: ZEEBE_BROKER_THREADS_CPUTHREADCOUNT
      value: "4"
    - name: ZEEBE_BROKER_THREADS_IOTHREADCOUNT
      value: "4"
    - name: ZEEBE_CLUSTER_SIZE
      value: "3"
    - name: ZEEBE_GATEWAY_MONITORING_ENABLED
      value: "true"
    - name: ZEEBE_LOG_APPENDER
      value: Stackdriver
    - name: ZEEBE_LOG_LEVEL
      value: debug
    - name: ZEEBE_PARTITIONS_COUNT
      value: "4"
    - name: ZEEBE_REPLICATION_FACTOR
      value: "3"
    image: camunda/zeebe:0.23.1
    imagePullPolicy: IfNotPresent
    livenessProbe:
      failureThreshold: 3
      httpGet:
        path: /ready
        port: 9600
        scheme: HTTP
      initialDelaySeconds: 900
      periodSeconds: 15
      successThreshold: 1
      timeoutSeconds: 1
    name: zeebe
    ports:
    - containerPort: 9600
      name: metrics
      protocol: TCP
    - containerPort: 26500
      name: gateway
      protocol: TCP
    - containerPort: 26501
      name: client
      protocol: TCP
    - containerPort: 26502
      name: management
      protocol: TCP
    - containerPort: 26503
      name: replication
      protocol: TCP
    - containerPort: 26504
      name: subscription
      protocol: TCP
    readinessProbe:
      failureThreshold: 3
      httpGet:
        path: /ready
        port: 9600
        scheme: HTTP
      initialDelaySeconds: 20
      periodSeconds: 5
      successThreshold: 1
      timeoutSeconds: 1
    resources:
      limits:
        cpu: "4"
        memory: 8Gi
      requests:
        cpu: 200m
        memory: 250Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /usr/local/zeebe/config/zeebe.cfg.yaml
      name: config
      subPath: zeebe.cfg.yaml
    - mountPath: /usr/local/zeebe/conf/zeebe.cfg.toml
      name: config
      subPath: zeebe.cfg.toml
    - mountPath: /usr/local/bin/startup.sh
      name: config
      subPath: startup.sh
    - mountPath: /usr/local/zeebe/data
      name: data
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-mntq5
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  hostname: zeebe-0
  nodeName: gke-ultratest-n1-highmem-8-pool-d05ffa7b-jd9k
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  subdomain: zeebe-broker-service
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - name: data
    persistentVolumeClaim:
      claimName: data-zeebe-0
  - configMap:
      defaultMode: 484
      name: zeebe-configmap
    name: config
  - name: default-token-mntq5
    secret:
      defaultMode: 420
      secretName: default-token-mntq5
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2020-05-19T11:35:02Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2020-05-19T13:12:30Z"
    message: 'containers with unready status: [zeebe]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2020-05-19T13:12:30Z"
    message: 'containers with unready status: [zeebe]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2020-05-19T11:35:02Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://5c3a8b749b33754d353f0a762cd20474af09cf693c11bdf31bb5430efa3c7355
    image: camunda/zeebe:0.23.1
    imageID: docker-pullable://camunda/zeebe@sha256:0a78fdb1533219678ff72332a26d25e689823e22a6b9b7f6e267866deface8f7
    lastState: {}
    name: zeebe
    ready: false
    restartCount: 0
    state:
      terminated:
        exitCode: 0
        finishedAt: null
        startedAt: null
  hostIP: 10.132.0.62
  phase: Running
  podIP: 10.28.12.3
  qosClass: Burstable
  startTime: "2020-05-19T11:35:02Z"

**Comment by hisImminence** [comment]
[stackdriver logs](https://console.cloud.google.com/logs/viewer?project=camunda-cloud-240911&minLogLevel=0&expandAll=false&timestamp=2020-05-19T13:03:54.182000000Z&customFacets=&limitCustomFacetWidth=true&advancedFilter=resource.type%3D%22k8s_container%22%0Aresource.labels.cluster_name%3D%22ultratest%22%0Aresource.labels.pod_name:%22zeebe-2%22%0Aresource.labels.namespace_name%3D%226122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe%22%0ANOT%20jsonPayload.message:%22snapshot%22%0A&interval=PT1H&scrollTimestamp=2020-05-19T11:36:03.973000000Z&dateRangeEnd=2020-05-19T12:58:59.793Z&dateRangeUnbound=backwardInTime)

**Comment by Zelldon** [comment]
**Observation**

We saw in the e2e test dashboard that the test were failing for a period of time and after that they are green again.

![e2e](https://user-images.githubusercontent.com/2758593/82332113-4a388500-99e5-11ea-8416-ef3184c0f99b.png)

For example we checked for `6122e231-7df7-4e0e-9f83-bbbd4eebbefc-zeebe` (M-Cluster) the failing intervals.

One interval was between 15.05.2020 18:30 till 16.05.2020 8:30 we were able to correlate these failing tests to leader changes on the Zeebe Cluster

![role-change-1505](https://user-images.githubusercontent.com/2758593/82332496-c4690980-99e5-11ea-8dc2-9bd748827a8f.png)

Between this time most of the requests were not forwarded to the brokers.

![receive-1505](https://user-images.githubusercontent.com/2758593/82332569-da76ca00-99e5-11ea-9f4b-30765ab35d34.png)

We were able to do the same for a second interval from 17.05.2020 22:00 'till 18.05.2020 02:45.

![role-change-1705](https://user-images.githubusercontent.com/2758593/82332835-30e40880-99e6-11ea-8069-b79a7291bbc7.png)

We checked elastic and also found no entry before 02:45, which confirms our assumption that requests are not getting to the broker.

![receive-1705](https://user-images.githubusercontent.com/2758593/82332923-51ac5e00-99e6-11ea-85f8-faead3b5392a.png)


Later we were able to check a currently failing e2e tests and requested the topology on all nodes. We saw then an incomplete topology on node 2 as posted here https://github.com/zeebe-io/zeebe/issues/4557#issuecomment-630808526

**Comment by hisImminence** [comment]
Data partitions are mounted on a VM, please contact @zoltan or me to get access if you need to. @Zelldon should have access already.

**Comment by Zelldon** [comment]
I think this is more critical than we thought. I have seen this also with the standalone gateway on our latest benchmark. 

```
[zell long-running/ ns:long-running]$ kgpo
NAME                                         READY   STATUS    RESTARTS   AGE
elasticsearch-master-0                       1/1     Running   0          20m
elasticsearch-master-1                       1/1     Running   0          20m
elasticsearch-master-2                       1/1     Running   0          20m
long-running-zeebe-0                         1/1     Running   0          20m
long-running-zeebe-1                         1/1     Running   0          20m
long-running-zeebe-2                         1/1     Running   0          20m
long-running-zeebe-gateway-877fc789b-5bjl7   1/1     Running   0          20m
starter-68fcfd4b9c-m4mxc                     1/1     Running   0          20m
worker-6ff67d4845-2krch                      1/1     Running   0          20m
worker-6ff67d4845-2r8gl                      1/1     Running   0          20m
worker-6ff67d4845-5g625                      1/1     Running   0          20m
worker-6ff67d4845-82lm8                      1/1     Running   0          20m
worker-6ff67d4845-bcslg                      1/1     Running   0          20m
worker-6ff67d4845-d6tmv                      1/1     Running   0          20m
worker-6ff67d4845-dpmt9                      1/1     Running   0          20m
worker-6ff67d4845-flfnr                      1/1     Running   0          20m
worker-6ff67d4845-sscjc                      1/1     Running   0          20m
worker-6ff67d4845-vq9zj                      1/1     Running   0          20m
worker-6ff67d4845-wzc5r                      1/1     Running   0          20m
worker-6ff67d4845-z7rwc                      1/1     Running   0          20m
[zell long-running/ ns:long-running]$ k exec -it long-running-zeebe-gateway-877fc789b-5bjl7 -- zbctl --insecure status
Cluster size: 3
Partitions count: 3
Replication factor: 3
Gateway version: 0.24.0-SNAPSHOT
Brokers:
  Broker 0 - long-running-zeebe-0.long-running-zeebe.long-running.svc.cluster.local:26501
    Version: 0.24.0-SNAPSHOT
    Partition 1 : Follower
    Partition 2 : Follower
    Partition 3 : Follower
  Broker 1 - long-running-zeebe-1.long-running-zeebe.long-running.svc.cluster.local:26501
    Version: 0.24.0-SNAPSHOT
    Partition 1 : Follower
    Partition 2 : Follower
    Partition 3 : Follower
  Broker 2 - long-running-zeebe-2.long-running-zeebe.long-running.svc.cluster.local:26501
    Version: 0.24.0-SNAPSHOT
    Partition 1 : Leader
```

This is quite problematic, because we just have here on leader where we can send requests to.

**Comment by npepinpe** [comment]
Let's work on this now then - as far as I'm concerned this is a pretty major issue if we cannot route requests properly to the brokers. The workaround of causing failover is not acceptable UX.

**Comment by Zelldon** [comment]
Ok I think I know what the issue is.

We introduced the term check in the TopologyManager, in order to not overwrite the topology with an old state. The problem is that we do not remove the term if we remove the Leader. This means if we removed the Leader, because Gossip detected the node as Dead then we remove everything excepted the last partition leader term. This is normally ok, since the next leader should have an greater term. But it can happen that this was just a network hickup and the old leader comes back as leader with the same term. In this case we will not add the Partition leader because we check for a greater term.

The problematic line is [here](https://github.com/zeebe-io/zeebe/blob/develop/gateway/src/main/java/io/zeebe/gateway/impl/broker/cluster/BrokerClusterStateImpl.java#L61):

```
  public void setPartitionLeader(final int partitionId, final int leaderId, final long term) {
    if (partitionLeaderTerms.getOrDefault(partitionId, -1L) < term) {
      partitionLeaders.put(partitionId, leaderId);
      partitionLeaderTerms.put(partitionId, Long.valueOf(term));
      final List<Integer> followers = partitionFollowers.get(partitionId);
      if (followers != null) {
        followers.removeIf(follower -> follower == leaderId);
      }
    }
  }
```

I observed that in my failing benchmark.

```
   
2020-05-20 11:49:54.618 CEST
2020-05-20 09:49:54.617 [GatewayTopologyManager] [gateway-scheduler-zb-actors-3] DEBUG io.zeebe.gateway - Received metadata change from Broker 0, partitions {1=FOLLOWER, 2=FOLLOWER, 3=FOLLOWER} and terms {}.
2020-05-20 11:50:01.647 CEST
2020-05-20 09:50:01.647 [GatewayTopologyManager] [gateway-scheduler-zb-actors-0] DEBUG io.zeebe.gateway - Received new broker BrokerInfo{nodeId=1, partitionsCount=3, clusterSize=3, replicationFactor=3, partitionRoles={1=FOLLOWER, 2=FOLLOWER, 3=FOLLOWER}, partitionLeaderTerms={}, version=0.24.0-SNAPSHOT}.
2020-05-20 11:50:01.647 CEST
2020-05-20 09:50:01.647 [GatewayTopologyManager] [gateway-scheduler-zb-actors-0] DEBUG io.zeebe.gateway - Received new broker BrokerInfo{nodeId=2, partitionsCount=3, clusterSize=3, replicationFactor=3, partitionRoles={1=LEADER, 2=LEADER, 3=LEADER}, partitionLeaderTerms={1=1, 2=1, 3=1}, version=0.24.0-SNAPSHOT}.
2020-05-20 11:50:04.612 CEST
2020-05-20 09:50:04.611 [GatewayTopologyManager] [gateway-scheduler-zb-actors-1] DEBUG io.zeebe.gateway - Received metadata change from Broker 2, partitions {1=LEADER, 2=LEADER, 3=LEADER} and terms {1=1, 2=1, 3=1}.
2020-05-20 11:53:08.751 CEST
2020-05-20 09:53:08.751 [GatewayTopologyManager] [gateway-scheduler-zb-actors-3] DEBUG io.zeebe.gateway - Received REACHABILITY_CHANGED for broker 2, do nothing.
2020-05-20 11:53:18.856 CEST
2020-05-20 09:53:18.855 [GatewayTopologyManager] [gateway-scheduler-zb-actors-0] DEBUG io.zeebe.gateway - Received broker was removed BrokerInfo{nodeId=2, partitionsCount=3, clusterSize=3, replicationFactor=3, partitionRoles={1=LEADER, 2=LEADER, 3=LEADER}, partitionLeaderTerms={1=1, 2=1, 3=1}, version=0.24.0-SNAPSHOT}.
2020-05-20 11:53:21.610 CEST
2020-05-20 09:53:21.609 [GatewayTopologyManager] [gateway-scheduler-zb-actors-3] DEBUG io.zeebe.gateway - Received new broker BrokerInfo{nodeId=2, partitionsCount=3, clusterSize=3, replicationFactor=3, partitionRoles={1=FOLLOWER, 2=LEADER, 3=LEADER}, partitionLeaderTerms={2=1, 3=1}, version=0.24.0-SNAPSHOT}.
2020-05-20 11:53:31.609 CEST
2020-05-20 09:53:31.609 [GatewayTopologyManager] [gateway-scheduler-zb-actors-3] DEBUG io.zeebe.gateway - Received metadata change from Broker 2, partitions {1=LEADER, 2=LEADER, 3=LEADER} and terms {1=3, 2=1, 3=1}.
```

I will verify this also in the logs from the cloud team to be sure.

**Comment by Zelldon** [comment]
Unfortunately the stackdriver logs are gone, but I'm pretty sure this was the issue. I created an PR which solves this #4564 . I will create PR's for backporting it to 0.22 and 0.23, when the review is done.

**Comment by strawhat5** [comment]
Great to see this! Saw this few weeks back on our standalone gateway cluster as well. Didn't report this thinking it might be our own network issue. Nice catch! :)

**Comment by shishirkhandelwal29** [comment]
Hi, Is this resolved? I am using 0.24.1 but still facing this issue.
